adds defer docs (#10079)
apollographql/apollo-client: docs/source/data/defer.mdx

---

> title: '@defer support'
> description: The @defer directive and Apollo Client

Reviewer: ```suggestion
title: 'Using the @defer directive in Apollo Client'
```

---

> > ⚠️ **Defer is a [Stage 1 GraphQL specification proposal](https://github.com/graphql/graphql-spec/blob/main/CONTRIBUTING.md#stage-1-proposal).** Before it reaches acceptance, it might still change in backward incompatible ways. If you have feedback on it, please let us know via [GitHub issues](https://github.com/apollographql/apollo-client/issues/new?assignees=&labels=&template=bug.md)

Reviewer 2: To this I'd add that `@defer` is now Stage 2 🎉

> Apollo Web Client supports [the `@defer` directive](https://github.com/graphql/graphql-wg/blob/main/rfcs/DeferStream.md), which allows receiving the fields of certain fragments of a query asynchronously.

Reviewer: Specify first supported version where indicated

> For instance consider a service where retrieving basic information from a user is fast, but retrieving the user's friends may take more time. In that case it would make sense to display the basic information in the UI as soon as they arrive, and display a loading indicator in the UI while waiting for the friends.

Reviewer: ```suggestion
For example, let's say we're building a social media application that can quickly fetch a user's basic profile information, but retrieving that user's friends takes longer. If we include _all_ of those fields in a single query, we want to be able to display the profile information as soon as it's available, instead of waiting for the friend fields to resolve.
```

> Applying the `@defer` directive to the fragment selecting the friends fields indicates that we are ready to receive them later, asynchronously:

Reviewer: ```suggestion
To achieve this, we can apply the `@defer` directive to an in-line fragment that contains all slow-resolving fields related to friend data:
```

```graphql
query PersonQuery($personId: ID!) {
  person(id: $personId) {
    # Basic fields (fast)
    id
    firstName
    lastName

>     # Friends (slow)
>     ... on User @defer {
>       friends {
>         id
>       }
>     }

Reviewer: ```suggestion
    # highlight-start
    # Friend fields (slower)
    ... on User @defer {
      friends {
        id
      }
    }
    # highlight-end
```

  }
}
```

> ## Code Generation

Reviewer: ```suggestion
## Using with code generation
```

> If you currently use [GraphQL Code Generator](https://www.the-guild.dev/graphql/codegen) for your CodeGen needs, be advised it doesn't yet support the use of the `@defer` directive in the code output.

Reviewer: ```suggestion
If you currently use [GraphQL Code Generator](https://www.the-guild.dev/graphql/codegen) for your codegen needs, note that it doesn't yet support the use of the `@defer` directive in the code output.
```

> You can refer to this [issue](https://github.com/dotansimha/graphql-code-generator/issues/7885) in the Github project for reference.

Reviewer: ```suggestion
For reference, refer to [this GitHub issue](https://github.com/dotansimha/graphql-code-generator/issues/7885).
```
